File: 0092/solution.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseBetween(
        self, head: Optional[ListNode], left: int, right: int
    ) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        curr, prev = dummy, None
        index = 0
        while index < left:
            prev = curr
            curr = curr.next
            index += 1

        left = prev
        left.next = None
        head = curr

        while index < right + 1:
            prev = curr
            curr = curr.next
            index += 1

        prev.next = None
        prev = curr

        while head:
            temp = head.next
            head.next = prev
            prev = head
            head = temp
        left.next = prev
        return dummy.next
        # time O(n)
        # space O(1)

File: 0092/solution.java

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode dummy = new ListNode(0, head);
        ListNode curr = dummy;
        while (right > 0) {
            curr = curr.next;
            right--;
        }
        ListNode prev = curr.next;
        curr.next = null;
        ListNode curr2 = dummy;
        while (left > 1) {
            curr2 = curr2.next;
            left--;
        }
        head = curr2.next;
        curr2.next = curr;
        while (head != null) {
            ListNode temp = head.next;
            head.next = prev;
            prev = head;
            head = temp;
        }
        return dummy.next;
    }
}